### ISSUE TITLE
Feature request: allow `predict_likelihood_parameters` for auto-regressive `QuantileRegression`

### ISSUE THREAD

**Issue by daidahao** [unit8co/darts#2912]
**Is your feature request related to a current problem? Please describe.**
Currently, sampling from torch-based, auto-regressive `QuantileRegression` models is slow and consumes a great amount of memory.
Allowing `predict_likelihood_parameters` for auto-regressive `QuantileRegression` could skip Monte Carlo sampling entirely and speed up forecasting. 

The issues has been explored previously in #2731, but I would like to re-open the issue here. I think a slight alternation to the solution might lead to better results.

**Describe proposed solution**
1. `model.predict()` allows `predict_likelihood_parameters=True` and `n>output_chunk_length` only when `num_samples=1` and `model.likelihood` is `QuantileRegression` (for now, other likelihoods could be explored in the future).
2. Inside `PLForecastingModule.predict_step()` when making auto-regressive, batch prediction, appends the **median ($q=0.5$)** prediction to the target.

This is also a common sampling approach with some of the latest deep learning models, such as Chronos-Bolt, TimesFM, etc.

**Describe potential alternatives**
We could also set `num_samples=len(quantiles)` for `model.predict()`. When appending the prediction, we first get new quantiles from such those samples and use those new quintiles as the new samples for target. However, this may add great complexity to the implementation since the samples might be spread across different batches.

**Additional context**
I tried replicating the experiment in #2731 (the best I could) and am getting better convergence than before. However, my model might not have the same hyper-parameters as @dennisbader, so please take that with a grain of salt.
- `input_chunk_length=24`
- `output_chunk_length=12`
- `quantiles=[0.1, 0.5, 0.9]`
- "forecast sampled" is plotted with `[0.1, 0.9]` quantiles.

On the other hand and fundamentally, this is a different sampling method than Monte Carlo. Divergence should be small but is expected over longer auto-regressive horizons.

<img width="1210" height="447" alt="Image" src="https://github.com/user-attachments/assets/867f164f-0d66-47be-815a-916c891559e4" />

**Comment by daidahao** [comment]
@dennisbader Let me know what you think of the approach here. :)

**Comment by dennisbader** [comment]
Thanks a lot @daidahao for taking another look at this issue :) 

From you description it sounds to me like the main difference from your approach compared to #2731 is computing the auto-regressive iterations consuming only the median forecasts (q=0.5) as input instead of forecasting each quantile path separately.

In general, I totally agree that it would be great to come up with a neat approach to reduce computational cost significantly.
What I see problematic is that using the non-monte carlo approach will most likely result in overconfident predictions (e.g. prediction interval width remains relatively constant with increasing forecast horizon) because it assumes that consumed input forecasts are actual historic measurements.

Can you try once increasing the dropout (a lot) in your model and see whether the monte-carlo approach results in wider auto-regressive forecast intervals compared to your approach?

**Comment by daidahao** [comment]
@dennisbader That's a great point. Using median forecast auto-regressively could lead to over-confident predictions. Following your suggestions, I re-run the experiments but with a large dropout 0.9 and even set `mc_dropout=True` for MC sampling. This is what I am getting:

<img width="1210" height="447" alt="Image" src="https://github.com/user-attachments/assets/48908cd2-07dc-4935-af31-536da39f5032" />

It seems that 1) MC sampling does not expand intervals along the horizon, and 2) the "median sampling" forecasts continues to get good convergence with MC sampling. To explain 1), it might be due to the fact that `AirPassengersDataset` is too small for uncertainty. 

In terms of computational costs, I see 2x speed-up (on MPS) even when I maximise the batch size for prediction. Interestingly, those speed-up disappear on CPU, likely due to less data transfer. But once I double the `num_samples` to 1000, the speed-up becomes 2x again.

**Word of Caution**
Even if "median sampling" forecasts diverge from MC sampling over the long horizon, it is fair to say neither is properly calibrated to the [0.1, 0.9] interval. Neither of the forecast intervals reflects the actual target distribution. It is expected that a user would pair the model with a conformal model to calibrate those intervals. That said, the current implementation of conformal model `predict()` overrides the `num_samples` and `predict_likelihood_parameters` options. 

We could apply the same logic here to conformal prediction so that even more computational costs can be saved. I will give you an update on that soon.

Here is my test script:

```python
import timeit

import numpy as np
from matplotlib import pyplot as plt

from darts.dataprocessing.transformers import Scaler
from darts.datasets import AirPassengersDataset
from darts.models import TiDEModel
from darts.utils.likelihood_models.torch import QuantileRegression

dropout = 0.9
pred_len = 48
quantiles = [0.1, 0.5, 0.9]

# Load data
series = AirPassengersDataset().load().astype(np.float32)
train, val = series[:-36], series[-36:]

# Scale data
scaler = Scaler()
train = scaler.fit_transform(train)
val = scaler.transform(val)
series = scaler.transform(series)

# Define and fit model
model = TiDEModel(
    input_chunk_length=24,
    output_chunk_length=12,
    batch_size=128,
    dropout=dropout,
    likelihood=QuantileRegression(quantiles),
    pl_trainer_kwargs={"accelerator": "mps"},
)
# Train on the whole series to align with #2731
model.fit(series, epochs=300)

# Make predictions
def fast_predict():
    """Predict with "median" sampling.

    This method appends the median prediction to the input window at each step.
    To implement it, I hacked into `PLForecastingModule.predict_step()` to make it
    append the median prediction to the batch output when the following conditions
    are all met:
    - `num_samples=1`,
    - `self.likelihood` is `QuantileRegression`,
    - `predict_likelihood_parameters=True`.
    Besides, several sanity checks are disabled to allow auto-regressive prediction.
    """
    return model.predict(
        n=pred_len,
        num_samples=1,
        predict_likelihood_parameters=True,
        verbose=False
    )

def slow_predict():
    """Predict with MC sampling."""
    return model.predict(
        n=pred_len,
        num_samples=500,
        mc_dropout=True,
        verbose=False,
        batch_size=2048,
    )

fast_timeit = timeit.timeit(fast_predict, number=10)
slow_timeit = timeit.timeit(slow_predict, number=10)
print(f"Fast predict time (10 runs): {fast_timeit:.2f} seconds")
print(f"Slow predict time (10 runs): {slow_timeit:.2f} seconds")
# Fast predict time (10 runs): 0.45 seconds
# Slow predict time (10 runs): 1.14 seconds

pred = fast_predict()
pred_sampled = slow_predict()


```

**Comment by daidahao** [comment]
Updates on conformal prediction:
I use a `ConformalQRModel` and apply the same logic to conformal prediction `.predict()` when the conditions are met:
- `model.likelihood` is `QuantileRegression`,
- `num_samples == 1`,
- `predict_likelihood_parameters=True`,
-  `quantiles == model.likelihood.quantiles`.

To implement the median sampling, I disable more sanity checks, override `cal_num_samples` to be 1 and `cal_predict_likelihood_parameters` to `True`, bypass `_apply_interval()` such that `np.quantile()` is never called. Instead, we now apply the `q_hat_` directly on top of the lower and upper quantiles, while the median quantile remains unchanged.

This is the result I get:

<img width="1218" height="447" alt="Image" src="https://github.com/user-attachments/assets/abcc1d77-11a0-4903-b2d4-a5ec4ba7ad38" />

With calibration applied, the median sampling and MC sampling are close to each other, despite their differences in the median. Median sampling shows a bit more over-confidence with a slightly narrower interval. (In some cases, a narrower interval might be preferred.)

To study the uncertainty, I also plotted the interval width (on the bottom) of both forecasts. We can see both widths increase with the forecast horizon. Again, we are seeing wider interval from MC sampling, especially after the first 12 months (when auto-regression starts).

In terms of speed, median sampling is 11x faster than MC sampling. Computation saves could be significant with a long horizon or large `cal_num_samples` / `num_samples`.

@dennisbader Looking forward to your thoughts on this!

```py
import timeit

import numpy as np

from darts.dataprocessing.transformers import Scaler
from darts.datasets import AirPassengersDataset
from darts.models import ConformalQRModel, TiDEModel
from darts.utils.likelihood_models.torch import QuantileRegression

dropout = 0.9
pred_len = 48
quantiles = [0.1, 0.5, 0.9]

# Load data
series = AirPassengersDataset().load().astype(np.float32)
train, val = series[:-36], series[-36:]

# Scale data
scaler = Scaler()
train = scaler.fit_transform(train)
val = scaler.transform(val)
series = scaler.transform(series)

# Define and fit model
model = TiDEModel(
    input_chunk_length=24,
    output_chunk_length=12,
    batch_size=128,
    dropout=dropout,
    likelihood=QuantileRegression(quantiles),
    pl_trainer_kwargs={"accelerator": "mps"},
)
# Train on the whole series to align with #2731
model.fit(series, epochs=300)

# Wrap with conformal prediction
conformal_model = ConformalQRModel(
    model=model,
    quantiles=quantiles,
    symmetric=False,
    cal_num_samples=500,
)

# Make predictions
def fast_predict():
    """Predict with "median" sampling.

    This method appends the median prediction to the input window at each step.
    To implement it, I hacked into `PLForecastingModule.predict_step()` to make it
    append the median prediction to the batch output when the following conditions
    are all met:
    - `num_samples=1`,
    - `self.likelihood` is `QuantileRegression`,
    - `predict_likelihood_parameters=True`.
    Besides, several sanity checks are disabled to allow auto-regressive prediction.
    """
    return conformal_model.predict(
        n=pred_len,
        num_samples=1,
        predict_likelihood_parameters=True,
        verbose=False
    )

def slow_predict():
    """Predict with MC sampling."""
    return conformal_model.predict(
        n=pred_len,
        num_samples=500,
        mc_dropout=True,
        verbose=False,
        batch_size=2048,
    )

fast_timeit = timeit.timeit(fast_predict, number=10)
slow_timeit = timeit.timeit(slow_predict, number=10)
print(f"Fast predict time (10 runs): {fast_timeit:.2f} seconds")
print(f"Slow predict time (10 runs): {slow_timeit:.2f} seconds")
# Fast predict time (10 runs): 2.27 seconds
# Slow predict time (10 runs): 26.60 seconds

pred = fast_predict()
pred_sampled = slow_predict()
```

**Comment by dennisbader** [comment]
Thanks @daidahao for the updates, I really like your analytical way of addressing the topics :) 🚀 

Would it be possible to open a draft PR or branch on your fork so that I could also play a bit around with it? I wonder how it will behave in other scenarios which are more complex than the air passengers scenario :)

It would be interesting to see how it looks on a case where the MC actually starts to widen up the interval significantly with increasing horizon.

> the median sampling and MC sampling are close to each other, despite their differences in the median.

Btw, this is related to an open issue #2830, as we currently do not adjust the median quantile in the conformal predictions (it always returns the actual model prediction). It's in our backlog but I couldn't yet find time to work on it 😅 (one good new though is that releasing the new Darts version at this very moment 🥳 )

**Comment by daidahao** [comment]
> (one good new though is that releasing the new Darts version at this very moment 🥳 )

🥳🥳🥳 That's great news! I am honoured to be part of the new release!

> Btw, this is related to an open issue https://github.com/unit8co/darts/issues/2830, as we currently do not adjust the median quantile in the conformal predictions (it always returns the actual model prediction). It's in our backlog but I couldn't yet find time to work on it 😅

I think the current implementation of returning the actual median is not inherently a bad thing. In many cases, we would want a median quantile that is optimised for accuracy, and not calibration. Happy to discuss this further though if you think the behaviour should change.

> Would it be possible to open a draft PR or branch on your fork so that I could also play a bit around with it? I wonder how it will behave in other scenarios which are more complex than the air passengers scenario :)

Yes, I will open a draft PR soon so we could both play with this proposed feature!

**Comment by dennisbader** [comment]
Hi @daidahao, i finally had some time to play around with your branch :) Thanks a lot again for all the experiments and the draft PR 🚀 It's great to see!

First, just some quick notes about the code snippet you shared:
- Conformal models should not be calibrated on the series (or a subsequence of the series) that the forecasting model was trained on. This will lead to over-confident calibrated prediction intervals which do not generalize to new data (especially in the AirPassengersDataset where the errors on the future are most likely not exchangeable with the errors on the training set). I hope we make it clear in our documentation about the CP approach (see [here](https://unit8co.github.io/darts/examples/23-Conformal-Prediction-examples.html#Disadvantages)).
- Also, due to the long forecast horizon we pick, there are really not many samples that can be used for calibration which might have another negative effect on the intervals.
- As you can see, the final calibrated prediction intervals are rather narrow which I would say is unexpected if we forecast that many steps into the future for the air passengers series. Of course this was just a dummy code example but I wanted to explain why the resulting intervals might look very similar for the autoregressive sampled- and new likelihood parameter approach :) 

Coming back to the experiments. I didn't use the conformal prediction due to the limitations on this example.
- The prediction intervals can really widen up for longer horizons for sampled predictions whereas the direct parameter predictions give a rather constant width. The sampled approach is the more realistic one, since uncertainty should increase with increasing horizon.
- Adding support for this comes with a couple of concerns: 
  - I fear that if we allow auto-regression with direct param predictions, that users would always use it without knowing about the limitations. 
  - The logic would only work for QuantileRegression which would break the "unified API". This would also introduce additional complexity of our conformal models.

All in all, I do have a couple of doubts about adding this, even though it can speed things up drastically. The complexity it introduces (in both downstream code, non-unified api, and user understanding of the limitations) is for me a bit too high at the moment to feel comfortable of adding it.

<details>

<summary>Code snippet</summary>

```py
import matplotlib.pyplot as plt
import numpy as np

from darts.dataprocessing.transformers import Scaler
from darts.datasets import AirPassengersDataset
from darts.models import TiDEModel
from darts.utils.likelihood_models.torch import QuantileRegression

dropout = 0.9
pred_len = 100
quantiles = [0.1, 0.5, 0.9]

# Load data
series = AirPassengersDataset().load().astype(np.float32)

# Scale data
series = Scaler().fit_transform(series)

# Define and fit model
model = TiDEModel(
    input_chunk_length=24,
    output_chunk_length=12,
    batch_size=128,
    dropout=dropout,
    random_state=42,
    likelihood=QuantileRegression(quantiles),
    pl_trainer_kwargs={"accelerator": "mps"},
)
# Train on the whole series to align with #2731
model.fit(series, epochs=300)

p_params = model.predict(
    n=pred_len,
    num_samples=1,
    predict_likelihood_parameters=True,
    verbose=False
)
p_sampled = model.predict(
    n=pred_len,
    num_samples=500,
    mc_dropout=True,
    verbose=False,
    batch_size=2048,
)

series[-48:].plot()
p_params.plot(label="Params Prediction")
p_sampled.plot(label="Sampled Prediction")
plt.show()
```

</details>


<img width="1267" height="548" alt="Image" src="https://github.com/user-attachments/assets/bef572f4-5dae-439d-bc8d-f2dd528c9098" />

**Comment by daidahao** [comment]
@dennisbader Thank you for your detailed responses! Having experimented with the median sampling and the Monte Carlo sampling methods on some larger datasets, I come to similar conclusions as yours:

For one, the default **MC sampling method produces more realistic forecasts** as the interval widens over the horizon. Median sampling (even after the conformal calibration) are over-confident and less realistic.

For the other, even though switching to median sampling speeds up forecasting quick a lot, **the speedups are less dramatic (about 2x-3x) on 1M+ point dataset**. This is because **most inference time is being spent on calibration, rather than model forward pass**. (P.S. It raises an interesting question, should we parallelise calibration to reduce overheads?)

Finally, just as you suggested, **implementing a different sampling method would break many APIs** while only providing benefits for the specific use cases of QR likelihood + QR conformal. It would probably introduce additional options for `predict()` and `historical_forecast()` that may cause confusion or over-simplification (without knowing the limitations).

However, I do think there are some practical benefits of using median sampling depending on the use cases:
1. Speed-ups can be significant when working with median sized datasets and median length horizons.
2. In our experiments, MC sampling requires a lot of samples (>>100) to produce smooth forecasts. Otherwise, there would be a lot of noises on the "edge" interval, say 1%-99%. Median sampling only requires one sample and is less noisy.
3. Median sampling is becoming more common in deep learning models and is analogous to greedy decoding in LLMs. Some optimisation strategies from LLMs can be applied here.

I will close the PR given that the limitations and implementation complexity of median sampling outweigh the benefits at this point. If anyone has better sampling strategies that could address those, feel free to tag me for a discussion :)

@dennisbader Thank you again for the comments.
